EIPS/eip-4750.md

```diff
@@ -0,0 +1,146 @@
+
+6. Sets `current_section_index` to `code_section_index` and `PC` to `0`, and execution continues in the called section.
+
+#### `RETF`
+
+1. Does not have immediate arguments.
+2. If data stack has less than `caller_stack_height + types[code_section_index].outputs`, execution results in exceptional halt.
+3. Pops nothing and pushes nothing to data stack.
+4. Pops an item from return stack and sets `current_section_index` and `PC` to values from this item.
+    4.1. If return stack is empty after this, execution halts with success.
+
+### Code Validation
+
+In addition to container format validation rules above, we extend code section validation rules (as defined in [EIP-3670](./eip-3670.md)).
+
+1. Code validation rules of EIP-3670 are applied to every code section.
+2. List of allowed *terminating instructions* in EIP-3670 is extended to include `RETF`. (*Note that `CALLF`, like other instructions with immediates, cannot be truncated.*)
+3. Code section is invalid in case an immediate argument of any `CALLF` is greater than or equal to the total number of code sections.
+4. `RJUMP` and `RJUMPI` immediate argument value (jump destination relative offset) validation:
+    4.1. Code section is invalid in case offset points to a position outside of section bounds.
+    4.2. Code section is invalid in case offset points to one of two bytes directly following `CALLF` instruction.
+
+### Execution
+
+1. Execution starts at the first byte of the first code section, and PC is set to 0.
+2. Return stack is initialized to contain one item: `(code_section_index = 0, offset = 0, stack_height = 0)`
+3. Destinations of jumps are allowed only to be inside current code section. `JUMP` and `JUMPI` result in exceptional halt when destination is outside of current section bounds.
+4. If any instruction would access a data stack item below `caller_stack_height`, execution results in exceptional halt. This rule replaces the old stack underflow check.
+
+#### Implications on the JUMPDEST analysis
+
+- Analysis is done separately for each section, i.e. output of entire analysis is `number_of_code_sections` lists of possible jump destinations.
+- Analysis is extended to consider 2 bytes directly following `CALLF` to be invalid jump destination
+
+## Rationale
+
+### `RETF` in the top frame ends execution vs exceptionally halts
+
+Alternative logic for executing `RETF` in the top frame could be to exceptionally halt execution, because there is arguably no caller for the starting function. This would mean that return stack is initialized as empty, and `RETF` exceptionally aborts when return stack is empty.
+
+We have decided in favor of always having at least one item in the return stack, becasue it allows to avoid having a special case for empty stack in the interprer loop stack underflow check. We keep the stack underflow rule general by having `caller_stack_height = 0` in the top frame.
```

Comment:
```suggestion
We have decided in favor of always having at least one item in the return stack, because it allows to avoid having a special case for empty stack in the interpreter loop stack underflow check. We keep the stack underflow rule general by having `caller_stack_height = 0` in the top frame.
```

Reply:
Oh come on, codespell should have caught this :)